### PR TITLE
adding heroku worker

### PR DESCRIPTION
**Pull request by samuelcolvin**
see #584.

Not yet ready.

### CODE REVIEW

**Comment by samuelcolvin** [comment]
before I go any further. @nvie @selwin how do we feel about dropping support for 2.6?

Obviously I can fix the currently failing issue but more broadly 2.6 is pretty ancient and haunted these days.

**Comment by proofit404** [comment]
I know that python 2.6 is old and hard to maintain in projects supports both python versions. But it used in production (in my case as well).

**Comment by selwin** [comment]
Python 2.6 has reached end of life in 2013 so I think we can justify not supporting it in the next release.

Thanks for working on this!

**Comment by samuelcolvin** [comment]
For me this is ready.

File: rq/worker.py [hunk 1]
```diff
@@ -708,3 +714,51 @@ def main_work_horse(self, *args, **kwargs):
     def execute_job(self, *args, **kwargs):
         """Execute job in same thread/process, do not fork()"""
         return self.perform_job(*args, **kwargs)
+
+
+class ShutDownImminentException(Exception):
```
Comment (selwin) [inline rq/worker.py]: Can we move this to `exceptions.py` to keep things tidy?

Comment (samuelcolvin) [inline rq/worker.py]: Absolutely, I'm away from my computer for the weekend, will do next week.

Comment (samuelcolvin) [inline rq/worker.py]: done.

**Comment by selwin** [comment]
@samuelcolvin thanks for the fix, but could you please resolve the conflicts so I can merge this in? It must have conflicted with another [PR](https://github.com/nvie/rq/pull/641) I merged in earlier.

It would be awesome if you could also add a little documentation about this in the `workers` section in the `gh-pages` branch (in a separate PR).

**Comment by samuelcolvin** [comment]
ok, rebased as requested. Should be ready to merge. I'll do a docs pull request when I have a chance.

**Comment by selwin** [comment]
@samuelcolvin thanks a lot!

**Comment by nvie** [comment]
Wow, this is a nice one! Thanks for adding this—this is going to be very useful for people on Heroku!

Sorry for my late (post-merge, even!) response here, but could we perhaps move this into a "contrib" style package? I.e. `from rq.contrib.heroku import HerokuWorker, ShutdownImminentException` ? That keeps it nicely outside of the main package, so non-Heroku users aren't bothered with it.

**Comment by spiliopoulos** [comment]
@selwin 

Hey @samuelcolvin, nice work! I have a quick question though. I have been working on another PR and while running the test suite I noticed some of your tests failing with
    
     AttributeError: 'frame' object has no attribute 'f_exc_traceback'

I took a quick look at your commits and I noticed that you make use of the following attributes of a frame on your signal handler:

     frame_properties = ['f_code', 'f_exc_traceback', 'f_exc_type', 'f_exc_value',                                                                                                                                                                                         
                         'f_lasti', 'f_lineno', 'f_locals', 'f_restricted', 'f_trace']

However, if I am not mistaken 3 of them have been removed in python3:
      f_exc_traceback, f_exc_type, f_exc_value

So I was wondering whether they should be removed and I was also very curious how did this pass from travis.

**Comment by spiliopoulos** [comment]
To answer my second question (why did this pass from Travis), it seems that as long as something raised an exception (even if it was not ShutDownImminentException) before the finished sentinel file was created the test was successful. In my case the exception was:

     AttributeError: 'frame' object has no attribute 'f_exc_traceback'

In any case I feel that the signal handler should be working correctly for all supported python versions. I could create a PR if you agree.

**Comment by samuelcolvin** [comment]
Sorry about the problem, I guess it's probably my responsibility to fix it but if you submit a pr that would be great too, otherwise I'll fix as soon as I can.

**Comment by spiliopoulos** [comment]
I'll see if I can work on it today.

**Comment by spiliopoulos** [comment]
PR #727

**Pull request merged**